### ISSUE TITLE
Extending Admin UI

### ISSUE THREAD

**Issue by plakhin** [conedevelopment/bazar#48]
What's the right way to add a page to Admin UI?
I want to implement basic CRUD for an Eloquent model.

**Comment by iamgergo** [comment]
You can publish the assets easily and implement any custom Vue/Inertia component.

However, I'm experimenting (for a while) with a new, easier way of extending the Admin UI.
Hopefully, I'll manage to do that in v0.3, so we'll have a full section of the documentation on how to extend the UI.

Thanks!

**Comment by iamgergo** [comment]
Returning here, it's been a while.

So, I was experimenting with a bunch of things, but my aim was to keep the concept of Inertia.js because it really gives a neat and fluid feeling for the admin UI. Keeping in mind, I found a solution that fairly similar that Laravel Nova does with scripts and styles.

Let's say you want to register automatically a custom script that holds your custom components and Inertia pages. It might come from your root application or from a package. To be able to do that, I'm planning to implement an `AssetRepository` that handles styles, scripts, and icons (for now), so you could do the following:

```php
// In a service provider

use Bazar\Support\Facades\Asset;

Asset::script('bazar-plugin', __DIR__.'/../public/bazar-plugin.js');
```

The script would be the compiled script, so it could be used instantly. Also, the `Asset` facade could symlink the registered assets, but if the user wants to overwrite them, they can be published (using the default publish functionality), and compile the modified scripts, so they can be used instead of the symlinked script. Also, symlinks would be working only, if there is no real file present on the given path.

From this point, we need to handle how to dynamically register components and pages to the Bazar Vue instance. I found the following (and quite not final) solution for this. You could do this in your registered script:

```js
import CustomComponent from './Components/Custom';

document.addEventListener('bazar:booting', event => {
    event.detail.Bazar.Vue.component('custom', CustomComponent);

    Object.assign(event.detail.Bazar.pages, {
         'bazar-plugin::ModelName/Index': import('./Pages/ModelName/Index'),
         'bazar-plugin::ModelName/Show': import('./Pages/ModelName/Show'),
         'bazar-plugin::ModelName/Create': import('./Pages/ModelName/Create'),
    });
});
```

> Note, imports and components could be both sync and asnyc. By using dynamic imports, WebPack will chunk the JS files automatically.

After registering the pages, you can give the name of the registered (and namespaced) name of the component:

```php
Component::render('bazar-plugin::ModelName/Index', [...]);
```

All works as expected. After more than two months of experimenting, I'm quite sure about this direction, but there are some notes to make:

- No need for the Vue compiler, we can use only the runtime build, it's a bit smaller, no feelable difference for me,
- When building a package, Vue needs to be included only in the dev environment, so the plugin scripts could be quite small,
- The Asset manager is not ready yet, the next big thing to build is going to be that, hopefully, I can ship that soon,
- To make the UI more easily extensible, I'm planning to implement a menu builder as well, so no need to hook into the `Sidebar.vue` component directly,
- ~~At the very moment, I find it better if we'll use our own "Inertia implementation", it would be a vastly simplified version of the current library and built only to run on Bazar,~~
- To write plugins easily, I'm planning to create a repository for a Bazar plugin skeleton that can be used easily by anyone,
- ~~I find it also possible, that we create a plugin interface that wraps all the things above, and provides an easy and fluent way to extend Bazar, not only the UI but the whole system.~~

This would be a quite big development, so probably it will take some time. So I'm happy to hear some thoughts or ideas!

**Comment by plakhin** [comment]
Looks good.
The only thing is 

>At the very moment, I find it better if we'll use our own "Inertia implementation", it would be a vastly simplified version of the current library and built only to run on Bazar,

may require too much additional time for maintenance.

**Comment by iamgergo** [comment]
Yes, I agree there are pros and cons to doing that, but we'll see.

Thanks.

**Comment by iamgergo** [comment]
Hi,

So, we introduced the features that I've talked about here: https://github.com/conedevelopment/bazar/issues/48#issuecomment-761765784, and also added many bugfixes and small improvements.

I released the `v0.5.0` version, so all the features should be available. I'm also planning to write a detailed release note, and of course, extend the documentation soon.

Thanks!
